tdms/src/surface_phasors.cpp

```diff
@@ -0,0 +1,189 @@
+  //array with each line being of the form Re(Ex) Im(Ex) Re(Ey) ... Im(Hz). Each line corresponds to the
+  //the vertex with the same line as in surface_phasors.surface_vertices
+  mwSize dims[3] = {n_surface_vertices, 6, f_ex_vector_size};
+  mx_surface_amplitudes = mxCreateNumericArray(3, (const mwSize *) dims, mxDOUBLE_CLASS, mxCOMPLEX);
+  surface_EHr = cast_matlab_3D_array(mxGetPr((mxArray *) mx_surface_amplitudes), dims[0], dims[1],
+                                     dims[2]);
+  surface_EHi = cast_matlab_3D_array(mxGetPi((mxArray *) mx_surface_amplitudes), dims[0], dims[1],
+                                     dims[2]);
+}
+
+void SurfacePhasors::normalise_surface(int frequency_index, complex<double> Enorm,
+                                       complex<double> Hnorm) {
+  complex<double> normalised_amplitude;
+
+  // E-field components
+  for (int vindex = 0; vindex < n_surface_vertices; vindex++) {
+    for (int i = 0; i < 3; i++) {
+      // again, if surface_EH was just one array of FullFieldSnapshots, we could use the multiply_by methods of that class rather than doing this awkward cast-then-recast. Alas, MATLAB woes
+      normalised_amplitude = complex<double>(surface_EHr[frequency_index][i][vindex],
+                                             surface_EHi[frequency_index][i][vindex]) /
+                             Enorm;
+      surface_EHr[frequency_index][i][vindex] = normalised_amplitude.real();
+      surface_EHi[frequency_index][i][vindex] = normalised_amplitude.imag();
+    }
+  }
+  // H-field components
+  for (int vindex = 0; vindex < n_surface_vertices; vindex++) {
+    for (int i = 3; i < 6; i++) {
+      normalised_amplitude = complex<double>(surface_EHr[frequency_index][i][vindex],
+                                             surface_EHi[frequency_index][i][vindex]) /
+                             Hnorm;
+      surface_EHr[frequency_index][i][vindex] = normalised_amplitude.real();
+      surface_EHi[frequency_index][i][vindex] = normalised_amplitude.imag();
+    }
+  }
+}
+
+void SurfacePhasors::extractPhasorsSurface(int frequency_index,
+                                           ElectricSplitField &E, MagneticSplitField &H, int n,
+                                           double omega, int Nt, int J_tot,
+                                           SimulationParameters &params, bool interpolate) {
+  int vindex;
+  FullFieldSnapshot F;
+  complex<double> phaseTermE, phaseTermH, cphaseTermE, cphaseTermH;
+
+  phaseTermE = fmod(omega * ((double) n) * params.dt, 2 * DCPI);
+  phaseTermH = fmod(omega * ((double) n + 0.5) * params.dt, 2 * DCPI);
+
+  cphaseTermH = exp(phaseTermH * IMAGINARY_UNIT) * 1. / ((double) Nt);
+  cphaseTermE = exp(phaseTermE * IMAGINARY_UNIT) * 1. / ((double) Nt);
+
+  //loop over every vertex in the list
+#pragma omp parallel default(shared) private(F, phaseTermE, phaseTermH, vindex)
```

Comment:
I don't have a specific suggestion for here, but in general, perhaps we should add comments about exactly how this is parallelised. I mean like: the parallelization strategy.

```c++
// here we can parallelise the calculation of foo as we loop over bar
```

Reply:
Also: did you test locally that with `OMP_NUM_THREADS=8` (or whatever) that we still pass all of the system tests?

Yes, I'm asking you to do a manual test 👎 . It's probably my fault because I didn't get to #89 yet.

Reply:
Have run these locally and looks like we're all good. Notable difference in speed of course (as expected), but the tests don't complain.